numeric_ranges_algorithms/numeric_ranges_algorithms.md

```diff
@@ -0,0 +1,484 @@
+
+### Algorithms that already have ranges versions
+
+The following numeric algorithms already have ranges versions in C++23.
+
+* `iota` already has a ranges version in C++23.  P3179R8 adds a parallel version.
+
+* `accumulate` performs operations sequentially.  Its parallel version is `reduce`, which we propose here.  The non-parallel version has been translated in C++23 into `fold_left`.
+
+### Algorithms with a ranges proposal in flight
+
+The following algorithm has a proposal in flight to add a ranges analog.
+
+* `partial_sum` performs operations sequentially.  Its parallel analogs are `inclusive_scan` and `exclusive_scan`, which we propose here.  For the non-parallel ranges version that returns a stateful binary operator, <a href="https://wg21.link/P2760R1">P2760R1</a> suggests a view instead of an algorithm.  <a href="https://wg21.link/P3351R2">P3351R2</a>, "`views::scan`," proposes this view.  P3351R2 is currently in SG9 (Ranges Study Group) review.
+
+### Algorithms that do not need ranges versions
+
+The following algorithms do not need ranges versions, since they can be replaced with existing views and ranges algorithms.
+
+#### `inner_product`
+
+`inner_product` performs operations sequentially.  It can be replaced with existing views and ranges algorithms, e.g., `ranges::fold_left(views::zip_transform(std::multiplies(), x, y), 0.0, std::plus())`.  P2214R2 argues against adding a ranges analog of `inner_product`, because it is less fundamental than other algorithms, and because it's not clear how to incorporate projections.
+
+#### `adjacent_difference`
+
+`adjacent_difference` can be replaced with a combination of  `adjacent_transform_view` (which was adopted in C++23) and `ranges::copy`.  In our experience, adjacent differences or their generalization are often used in combination with other ranges.  For example, finite-difference methods for solving time-dependent differential equations may need to add together multiple ranges, each of which is an adjacent difference possibly composed with other functions.  One could represent the spatial finite difference scheme for a partial differential equation as a weighted sum of adjacent differences in each spatial degree of freedom.  Thus, for us a view would make more sense than a "terminal" algorithm.  The actual algorithm is a transform or copy from a complicated view into an output range.

```

Comment:
Are we sure though that a straightforward generic parallelization of `copy` or `transform` will correctly work for the use cases of `adjacent_difference` (or its generalizations)? If there are non-trivial data dependencies, I'm afraid it might not work properly and/or not efficiently with execution policies.
In other words, would not support for execution policies necessitate or at least make reasonable the dedicated function for this?

Reply:
> the burden is on us if we want to introduce algorithms other than reduce, inclusive_scan, and exclusive_scan. We have to show that we need the algorithm, for example because it's too hard for users to implement correctly and efficiently (like reduce and scans).

I agree with this assessment. Essentially, I would like us to double-check for `adjacent_difference` that a "naive" implementation composed from the existing algorithms and views would not have data races or performance problems.

Reply:
@akukanov Thanks! : - ) 

Btw, `adjacent_transform_view` has the same issue as `transform_view`, that it might not be trivially copyable for some lambdas that are trivally copyable.  However, if `binary_op` is just `std::minus`, there's no problem.